File: SDKs/flutter/lib/src/ivx_deep_links.dart
// Copyright (c) 2026 Intelli-verse-X
// MIT License — see LICENSE in the project root.

import 'dart:async';

/// Parsed result of a deep link URL.
class DeepLinkResult {
  final bool matched;
  final String scheme;
  final String host;
  final String route;
  final Map<String, String> params;
  final String raw;

  const DeepLinkResult({
    required this.matched,
    this.scheme = '',
    this.host = '',
    this.route = '',
    this.params = const {},
    this.raw = '',
  });
}

/// Handler signature for registered deep link routes.
typedef DeepLinkHandler = void Function(
    Map<String, String> params, DeepLinkResult result);

/// Lightweight deep link parser and dispatcher.
///
/// Parses URLs in the format `{scheme}://{host}/{route}?key=value`
/// and fires registered handlers for matching routes.
class IVXDeepLinks {
  static IVXDeepLinks? _instance;

  String _scheme = '';
  String _host = '';
  bool _initialized = false;

  final Map<String, List<DeepLinkHandler>> _handlers = {};
  final StreamController<DeepLinkResult> _linkController =
      StreamController<DeepLinkResult>.broadcast();

  IVXDeepLinks._();

  /// Shared singleton instance.
  static IVXDeepLinks get instance => _instance ??= IVXDeepLinks._();

  /// Stream of all successfully matched deep links.
  Stream<DeepLinkResult> get onDeepLink => _linkController.stream;

  /// Whether [initialize] has been called.
  bool get isInitialized => _initialized;

  /// Configure the expected scheme and host.
  void initialize({required String scheme, required String host}) {
    _scheme = scheme;
    _host = host;
    _initialized = true;
  }

  /// Parse [url] and dispatch to registered handlers.
  DeepLinkResult handleUrl(String url) {
    final result = _parse(url);
    if (result.matched) {
      _dispatch(result);
      _linkController.add(result);
    }
    return result;
  }

  /// Register a [handler] for a specific [route].
  void registerHandler(String route, DeepLinkHandler handler) {
    _handlers.putIfAbsent(route, () => []);
    _handlers[route]!.add(handler);
  }

  /// Remove a previously registered [handler] from a [route].
  void removeHandler(String route, DeepLinkHandler handler) {
    _handlers[route]?.remove(handler);
  }

  /// Remove all handlers, or only those for a specific [route].
  void removeAllHandlers([String? route]) {
    if (route != null) {
      _handlers.remove(route);
    } else {
      _handlers.clear();
    }
  }

  /// Release resources. After calling this, [instance] returns a fresh object.
  void dispose() {
    _linkController.close();
    _handlers.clear();
    _instance = null;
  }

  DeepLinkResult _parse(String url) {
    const empty = DeepLinkResult(matched: false);

    final schemeEnd = url.indexOf('://');
    if (schemeEnd == -1) return DeepLinkResult(matched: false, raw: url);

    final scheme = url.substring(0, schemeEnd);
    final rest = url.substring(schemeEnd + 3);

    final pathStart = rest.indexOf('/');
    final host = pathStart == -1 ? rest : rest.substring(0, pathStart);

    if (_initialized && (scheme != _scheme || host != _host)) {
      return DeepLinkResult(matched: false, raw: url);
    }

    final pathAndQuery = pathStart == -1 ? '' : rest.substring(pathStart + 1);
    final queryStart = pathAndQuery.indexOf('?');
    final route =
        queryStart == -1 ? pathAndQuery : pathAndQuery.substring(0, queryStart);
    final queryString =
        queryStart == -1 ? '' : pathAndQuery.substring(queryStart + 1);

    final params = <String, String>{};
    if (queryString.isNotEmpty) {
      for (final pair in queryString.split('&')) {
        final eqIdx = pair.indexOf('=');
        if (eqIdx == -1) {
          params[Uri.decodeComponent(pair)] = '';
        } else {
          params[Uri.decodeComponent(pair.substring(0, eqIdx))] =
              Uri.decodeComponent(pair.substring(eqIdx + 1));
        }
      }
    }

    return DeepLinkResult(
      matched: true,
      scheme: scheme,
      host: host,
      route: route,
      params: params,
      raw: url,
    );
  }

  void _dispatch(DeepLinkResult result) {
    final handlers = _handlers[result.route];
    if (handlers == null) return;
    for (final handler in List<DeepLinkHandler>.of(handlers)) {
      try {
        handler(result.params, result);
      } catch (_) {
        // Handler errors are silently swallowed to avoid cascading failures.
      }
    }
  }
}

File: SDKs/javascript/src/IVXDeepLinks.ts
// Copyright (c) 2026 Intelli-verse-X
// MIT License — see LICENSE in the project root.

export interface DeepLinkConfig {
  scheme: string;
  host: string;
}

export interface DeepLinkResult {
  matched: boolean;
  scheme: string;
  host: string;
  route: string;
  params: Record<string, string>;
  raw: string;
}

export type DeepLinkHandler = (params: Record<string, string>, result: DeepLinkResult) => void;

export class IVXDeepLinks {
  private static _instance: IVXDeepLinks | null = null;
  private _scheme = '';
  private _host = '';
  private _initialized = false;
  private _handlers = new Map<string, Set<DeepLinkHandler>>();

  private constructor() {}

  static getInstance(): IVXDeepLinks {
    if (!IVXDeepLinks._instance) {
      IVXDeepLinks._instance = new IVXDeepLinks();
    }
    return IVXDeepLinks._instance;
  }

  static initialize(config: DeepLinkConfig): void {
    const inst = IVXDeepLinks.getInstance();
    inst._scheme = config.scheme;
    inst._host = config.host;
    inst._initialized = true;
  }

  static handleUrl(url: string): DeepLinkResult {
    const inst = IVXDeepLinks.getInstance();
    const result = inst._parse(url);
    if (result.matched) {
      inst._dispatch(result);
    }
    return result;
  }

  static registerHandler(route: string, handler: DeepLinkHandler): void {
    const inst = IVXDeepLinks.getInstance();
    if (!inst._handlers.has(route)) {
      inst._handlers.set(route, new Set());
    }
    inst._handlers.get(route)!.add(handler);
  }

  static removeHandler(route: string, handler: DeepLinkHandler): void {
    const inst = IVXDeepLinks.getInstance();
    inst._handlers.get(route)?.delete(handler);
  }

  static removeAllHandlers(route?: string): void {
    const inst = IVXDeepLinks.getInstance();
    if (route) {
      inst._handlers.delete(route);
    } else {
      inst._handlers.clear();
    }
  }

  static get isInitialized(): boolean {
    return IVXDeepLinks.getInstance()._initialized;
  }

  private _parse(url: string): DeepLinkResult {
    const empty: DeepLinkResult = {
      matched: false,
      scheme: '',
      host: '',
      route: '',
      params: {},
      raw: url,
    };

    const schemeEnd = url.indexOf('://');
    if (schemeEnd === -1) return empty;

    const scheme = url.substring(0, schemeEnd);
    const rest = url.substring(schemeEnd + 3);

    const pathStart = rest.indexOf('/');
    const host = pathStart === -1 ? rest : rest.substring(0, pathStart);

    if (this._initialized && (scheme !== this._scheme || host !== this._host)) {
      return empty;
    }

    let pathAndQuery = pathStart === -1 ? '' : rest.substring(pathStart + 1);
    const queryStart = pathAndQuery.indexOf('?');
    const route = queryStart === -1 ? pathAndQuery : pathAndQuery.substring(0, queryStart);
    const queryString = queryStart === -1 ? '' : pathAndQuery.substring(queryStart + 1);

    const params: Record<string, string> = {};
    if (queryString) {
      for (const pair of queryString.split('&')) {
        const eqIdx = pair.indexOf('=');
        if (eqIdx === -1) {
          params[decodeURIComponent(pair)] = '';
        } else {
          params[decodeURIComponent(pair.substring(0, eqIdx))] =
            decodeURIComponent(pair.substring(eqIdx + 1));
        }
      }
    }

    return { matched: true, scheme, host, route, params, raw: url };
  }

  private _dispatch(result: DeepLinkResult): void {
    const handlers = this._handlers.get(result.route);
    if (handlers) {
      for (const handler of handlers) {
        try {
          handler(result.params, result);
        } catch (_) {
          // Handler errors are silently swallowed to avoid cascading failures.
        }
      }
    }
  }
}
